File: backend/src/scanner.ts
import { chromium, Page } from 'playwright';
import { Cookie } from 'playwright-core';

export interface RawScanData {
  cookies: Cookie[];
  scripts: { src: string }[];
  html: string;
}

// Function to try and accept common cookie banners
async function acceptCookieBanners(page: Page) {
  const selectors = [
    'button:has-text("Accept all")',
    'button:has-text("Accept")',
    'button:has-text("Allow all")',
    'button:has-text("Allow")',
    'button:has-text("I agree")',
    'button:has-text("OK")',
    '[id*="consent"]',
    '[class*="consent"]',
    '[id*="cookie"]',
    '[class*="cookie"]',
  ];

  for (const selector of selectors) {
    try {
      // Use page.locator which is the modern way to select elements
      const element = page.locator(selector).first();
      if (await element.isVisible()) {
        await element.click({ timeout: 2000 });
        console.log(`Clicked on a consent banner element: ${selector}`);
        // Wait a moment for the banner to disappear and cookies to be set
        await page.waitForTimeout(1000); 
        return; // Assume one banner is enough
      }
    } catch (e) {
      // Element not found, not visible, or not clickable, continue to the next
    }
  }
  console.log("No common cookie banners found to accept.");
}

export async function scanWebsite(url: string): Promise<RawScanData> {
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    // Emulate a user agent to avoid simple bot detection
    userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
  });
  const page = await context.newPage();

  try {
    await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 25000 });
    
    // Attempt to accept cookie banners to reveal all cookies and scripts
    await acceptCookieBanners(page);

    // Re-wait for network to be idle after potential banner interaction
    await page.waitForLoadState('networkidle', { timeout: 15000 });

    const cookies = await context.cookies();
    const html = await page.content();

    const scripts = await page.evaluate(() => {
      return Array.from(document.querySelectorAll('script[src]')).map(script => ({
        src: (script as HTMLScriptElement).src
      }));
    });

    return { cookies, scripts, html };

  } finally {
    await browser.close();
  }
}

File: scanner.py
from playwright.sync_api import sync_playwright
import re

def scan_website(url: str):
    """
    Scans a given website and extracts cookies, scripts, and detects
    basic privacy compliance issues such as missing Secure cookies or
    consent banners.
    """
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context()
        page = context.new_page()

        # Navigate to URL
        page.goto(url, wait_until="networkidle", timeout=45000)

        # Extract cookies and script sources
        cookies = context.cookies()
        scripts = [
            s.get_attribute("src")
            for s in page.query_selector_all("script")
            if s.get_attribute("src")
        ]

        # Check for consent banner
        html_content = page.content().lower()
        consent_banner = bool(re.search(r"(cookie.*consent|accept cookies)", html_content))

        # Detect insecure cookies (missing Secure/HttpOnly/SameSite)
        insecure_cookies = []
        for c in cookies:
            if not c.get("secure") or c.get("sameSite") == "None":
                insecure_cookies.append(c)

        result = {
            "url": url,
            "cookies": cookies,
            "scripts": scripts,
            "consent_banner": consent_banner,
            "violations": {
                "missing_consent_banner": not consent_banner,
                "insecure_cookies": len(insecure_cookies) > 0,
            },
            "insecure_cookie_count": len(insecure_cookies)
        }

        browser.close()
        return result


if __name__ == "__main__":
    # Local test
    import json
    print(json.dumps(scan_website("https://example.com"), indent=2))
